crates/sui-keys/src/external.rs

```diff
@@ -0,0 +1,916 @@
+                        "Cannot write aliases to file: {}",
+                        path.with_extension("aliases").display()
+                    )
+                })?;
+            Ok(())
+        } else {
+            Err(anyhow!("Path is not set for External keystore"))
+        }
+    }
+
+    pub async fn save_keystore(&self) -> Result<(), Error> {
+        if let Some(path) = &self.path {
+            let store: String = serde_json::to_string_pretty(&self.keys)
+                .map_err(|e| anyhow!("Serialization error: {}", e))?;
+
+            let keystore_path = path.clone();
+            tokio::task::spawn_blocking(move || std::fs::write(keystore_path, store))
+                .await?
+                .with_context(|| format!("Cannot write keystore to file: {}", path.display()))?;
+            Ok(())
+        } else {
+            Err(anyhow!("Path is not set for External keystore"))
+        }
+    }
+
+    pub async fn save(&self) -> Result<(), Error> {
+        self.save_aliases().await?;
+        self.save_keystore().await?;
+        Ok(())
+    }
+}
+
+impl Serialize for External {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(
+            self.path
+                .as_ref()
+                .unwrap_or(&PathBuf::default())
+                .to_str()
+                .unwrap_or(""),
+        )
+    }
+}
+
+impl<'de> Deserialize<'de> for External {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        use serde::de::Error;
+        External::new(&PathBuf::from(String::deserialize(deserializer)?)).map_err(D::Error::custom)
+    }
+}
+
+#[async_trait]
+impl AccountKeystore for External {
+    async fn generate(
+        &mut self,
+        alias: Option<String>,
+        opts: GenerateOptions,
+    ) -> Result<(SuiAddress, PublicKey, SignatureScheme), Error> {
+        let GenerateOptions::ExternalSigner(signer) = opts else {
+            return Err(anyhow!("Signer must be provided for external keys."));
+        };
+
+        let res = self.exec(&signer, "create_key", json![null]).await?;
+        let ExternalKey { key_id, public_key } = serde_json::from_value(res)
+            .map_err(|e| anyhow!("Failed to parse key response: {}", e))?;
+        let public_key = PublicKey::decode_base64(&public_key)?;
+        let address: SuiAddress = (&public_key).into();
+
+        self.keys.insert(
+            address,
+            StoredKey {
+                public_key: public_key.clone(),
+                signer: signer.clone(),
+                key_id: key_id.to_string(),
+            },
+        );
+
+        self.aliases.insert(
+            address,
+            Alias {
+                alias: self.create_alias(alias)?,
+                public_key_base64: public_key.encode_base64(),
+            },
+        );
+
+        Ok((address, public_key, ED25519))
```

Comment:
external signers might have other schemes 
might want to replace `ED25519` with `public_key.scheme()`